### ISSUE TITLE
Tree reset after each test

### ISSUE THREAD

**Issue by myin142** [bitwes/Gut#251]
When each test is called, will the whole tree be reset to the original structure? For example if I add a custom scene to the test tree in a `before_each` and do some changes to it. Will it be reset properly if a new test starts?

Because right now I have a problem that the exact same test fails if it is called the second time. The first test always succeeds but the second always fails. I checked the children count with `get_child_count` and it seems to be the same on each test, so I assume the scene is cleared after each test. But it seems like something is not cleaned up correctly.

Btw. I using the `get_tree().input_event()` to send inputs to move the player. The first time is moves but the second time the player just stands still. I checked the `_unhandled_input` function of the classes and it seems like on the second test there are no events.

**Comment by bitwes** [comment]
GUT doesn't do anything to the tree except to manage its own children.  Nothing in the tree "above" GUT or at the same level should be altered.  GUT will remove children and free objects added with the various auto-free methods after each test (after `after_each`).  At the end of a test, the test script (inheriting from `res://addons/gut/test.gd`) is removed from the tree but is not freed until the end of the test run.

There might be some timing issue with `input_event`.  Try adding a small pause at the start of  your `before_each` or in `after_each`.  
```
yield(yield_for(.1), YIELD)
```
If this works then this should be considered a workaround and this should be looked into further to see if something can be done to make this easier.

If that doesn't work, see if you can come up with simple example and post it here.

**Comment by bitwes** [comment]
I got curious about testing inputs in general. I think I was able to replicate your issue.  I intermittently got a failing 2nd test.  The tests look like:
```gdscript
func test_moves_right_5_after_1_second_press():
	var c = add_child_autofree(InputActionCharacter.instance())

	var a = InputEventAction.new()
	a.action = "ia_right"
	a.pressed = true

	get_tree().input_event(a)
	simulate(c, 2, .5)
	assert_eq(c.position.x, 100.0)

func test_moves_left_5_after_1_second_press():
	var c = add_child_autofree(InputActionCharacter.instance())

	var a = InputEventAction.new()
	a.action = "ia_left"
	a.pressed = true

	get_tree().input_event(a)
	simulate(c, 2, .5)
	assert_eq(c.position.x, -100.0)
```
I found that the tests would pass consistently if I called `_input` on my character instance instead of using `get_tree().input_event(a)`

In general, you shouldn't be calling "underscore" methods in your tests, I think this should be the recommended approach.

**Comment by myin142** [comment]
I tried to wait after each test but it seems to cause script errors. Some values are now null after waiting. I also noticed that it also causes tests of other classes to fail.
Using `_input` is not really an option for me. It is not certain which node will be processing the input. Right now the state machine is handling it, but if I change it in the future, I would have to update all my tests.

Do you still need a simple example or is yours enough?

**Comment by bitwes** [comment]
Related to #171.  We need a standard approach for mocking input.  I'm not sure what form this would take.  [This comment](https://github.com/bitwes/Gut/issues/171#issuecomment-700254936) covers some approaches.

**Comment by bitwes** [comment]
#171 has been implemented and slated for 7.3.0.  There are no other issues in here that are not covered by that issue so I'm closing this out.
